packages/@aws-cdk/aws-ecr/lib/public-repository.ts

```diff
@@ -0,0 +1,428 @@
+  readonly publicRepositoryArn: string;
+
+  /**
+   * Grant the given identity permissions to push images to this repository.
+   */
+  grantPush(grantee: iam.IGrantable): iam.Grant;
+}
+
+/**
+ * Base class for ECR Public Repository. Reused between imported repositories and owned repositories.
+ */
+export abstract class PublicRepositoryBase extends Resource implements IPublicRepository {
+
+  protected static readonly PUBLIC_REPO_REGION = 'us-east-1';
+
+  // From https://docs.aws.amazon.com/AmazonECR/latest/public/public-repository-policy-examples.html
+  private static readonly PUSH_IAM_ACTIONS = [
+    'ecr-public:BatchCheckLayerAvailability',
+    'ecr-public:CompleteLayerUpload',
+    'ecr-public:InitiateLayerUpload',
+    'ecr-public:PutImage',
+    'ecr-public:UploadLayerPart',
+  ];
+
+  public abstract readonly publicRepositoryName: string;
+  public abstract readonly publicRepositoryArn: string;
+
+  public abstract addToResourcePolicy(statement: iam.PolicyStatement): iam.AddToResourcePolicyResult;
+
+  public onCloudTrailEvent(id: string, options: events.OnEventOptions = {}): events.Rule {
+    // See https://docs.aws.amazon.com/AmazonECR/latest/public/logging-using-cloudtrail.html
+    if (Stack.of(this).region !== PublicRepositoryBase.PUBLIC_REPO_REGION) {
+      throw new Error(`CloudWatch event rule for ECR Public Repository can only be created in ${PublicRepositoryBase.PUBLIC_REPO_REGION}.`);
+    }
+
+    const rule = new events.Rule(this, id, options);
+    rule.addTarget(options.target);
+    rule.addEventPattern({
+      source: ['aws.ecr-public'],
+      detailType: ['AWS API Call via CloudTrail'],
+      detail: {
+        requestParameters: {
+          repositoryName: [this.publicRepositoryName],
+        },
+      },
+    });
+    return rule;
+  }
+
+  public onCloudTrailImagePushed(id: string, options: OnCloudTrailImagePushedOptions = {}): events.Rule {
+    const rule = this.onCloudTrailEvent(id, options);
+    rule.addEventPattern({
+      detail: {
+        eventName: ['PutImage'],
+        requestParameters: {
+          imageTag: options.imageTag ? [options.imageTag] : undefined,
+        },
+      },
+    });
+    return rule;
+  }
+
+  public onEvent(id: string, options: events.OnEventOptions = {}) {
+    if (Stack.of(this).region !== PublicRepositoryBase.PUBLIC_REPO_REGION) {
+      throw new Error(`CloudWatch event rule for ECR Public Repository can only be created in ${PublicRepositoryBase.PUBLIC_REPO_REGION}.`);
+    }
+    const rule = new events.Rule(this, id, options);
+    rule.addEventPattern({
+      source: ['aws.ecr-public'],
+      resources: [this.publicRepositoryArn],
+    });
+    rule.addTarget(options.target);
+    return rule;
+  }
+
+  public grant(grantee: iam.IGrantable, ...actions: string[]) {
+    return iam.Grant.addToPrincipalAndResource({
+      grantee,
+      actions,
+      resourceArns: [this.publicRepositoryArn],
+      resourceSelfArns: [],
+      resource: this,
+    });
+  }
```

Comment:
Can be shared with `RepositoryBase`

Reply:
I cannot make `grant`method to be defined in parent class because `PublicRepositoryBase` class already extends `Resource` class, and Typescript does not support multiple inheritance.

If you have alternative ideas let me know.